// setenv/src/setEnv.js
/* jsling node: true */
var exec = require('child_process').exec;


// constants
var UNIX_BASH_ENV = "~/.bashrc";
var UNIX_ZSH_ENV = "~/.zshrc";
var UNIX_CSH_ENV = "~/.login";
var UNIX_OTHER_SHELL_ENV = "~/.profile";


// global variables
var gVerbosity = false;



function executeScript( scriptToExecute ){
	exec( scriptToExecute,
		function( error, stdout, stderr ){
			if( gVerbosity ){
				console.log('stdout: ' + stdout);
				console.log('stderr: ' + stderr);
			} // if
			
			if (error !== null) {
				console.log('Error: ' + error);
			} // if
		}); // exec()
}


function createScriptForLinux(fileName, varaibleName, variableValue){
	var saveOrgFileScript = "cp " + fileName + " " + fileName + "_org";
	var editFileScript = "sed -i '/" + varaibleName + "/d;$ a\\export " + varaibleName + "=" + variableValue + "' " + fileName;
	var script = saveOrgFileScript + " && " + editFileScript;
	
	return script;
}


function setEnvLinux( name, value ){
	var shell = process.env.SHELL;
	var script;
	
	if( shell.indexOf('bash') !== -1 ){
		script = createScriptForLinux(UNIX_BASH_ENV, name, value);
	} else if( shell.indexOf('zsh') !== -1 ){
		script = createScriptForLinux(UNIX_ZSH_ENV, name, value);
	} else if( shell.indexOf('csh') !== -1 ){
		script = createScriptForLinux(UNIX_CSH_ENV, name, value);
	} else {
		script = createScriptForLinux(UNIX_OTHER_SHELL_ENV, name, value);
	}
	
	executeScript( script );
}


function setEnvWindows( name, value ){
	var script = 'SetX ' + name + ' ' + value;
	executeScript( script );
}


function setEnv( name, opt_value ){
	var script = "";
	opt_value = opt_value || name;
	
	
	// set environment variable for current process
	process.env[name] = opt_value;
	
	
	// permanently set environment variable
	if( process.platform === 'win32' ){
		setEnvWindows( name, opt_value );
	} else {
		setEnvLinux( name, opt_value );
	} // if
}


// publish 
module.exports.verbosity = gVerbosity;
module.exports.set = setEnv;
